Softeq.XToolkit.DefaultAuthorization/SecuredHttpServiceGate.cs

```diff
@@ -107,39 +83,51 @@ protected virtual void HandleInvalidResponse(HttpResponse response)
             }
         }
 
-        private async Task<ExecutionStatus> RetrieveSessionAsync()
+        private async Task<HttpResponse> RefreshTokenAndExecuteAsync(
+            HttpRequest request,
+            int timeout,
+            HttpRequestPriority priority,
+            params HttpStatusCode[] ignoreErrorCodes)
         {
-            var deferral = CreateSessionRetrievalDeferral();
-
-            deferral.Begin();
-
-            var result = await _sessionApiService.RefreshTokenAsync().ConfigureAwait(false);
+            await RefreshTokenAsync().ConfigureAwait(false);
 
-            deferral.Complete(result);
-
-            return result;
-        }
+            request.WithCredentials(_tokenManager);
+            var response = await _client.ExecuteAsStringResponseAsync(request, priority, timeout).ConfigureAwait(false);
+            if (response != null && response.StatusCode == HttpStatusCode.Unauthorized)
+            {
+                throw new InvalidSessionException("Got 401 status even after refreshing access token");
+            }
 
-        private ForegroundTaskDeferral<ExecutionStatus> CreateSessionRetrievalDeferral()
-        {
-            _sessionRetrievalDeferral = new ForegroundTaskDeferral<ExecutionStatus>();
+            if (response == null || !response.IsSuccessful && !ignoreErrorCodes.Contains(response.StatusCode))
+            {
+                HandleInvalidResponse(response);
+            }
 
-            return _sessionRetrievalDeferral;
+            return response;
         }
 
-        private async Task<ExecutionStatus?> EnsureNoSessionRetrievalIsRunningAsync()
+        private async Task RefreshTokenAsync()
         {
-            if (!_sessionRetrievalDeferral.IsInProgress)
+            ExecutionStatus refreshingTokenResult;
+            if (_refreshingTokenTask == null)
             {
-                return null;
+                _refreshingTokenTask = _sessionApiService.RefreshTokenAsync();
```

Comment:
Please revert `deferral` back
The current solution for multiple parallel requests will be trying refresh token for each request.